File: src/main/kotlin/dev/jefy/connectpro/portfolio/applicaion/query/PortfolioQueryImpl.java
package dev.jefy.connectpro.portfolio.applicaion.query;


import org.jspecify.annotations.NonNull;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.Assert;

import java.util.List;
import java.util.function.Function;

import dev.jefy.connectpro.engagement.EngagementClient;
import dev.jefy.connectpro.management.ManagementClient;
import dev.jefy.connectpro.management.appliacaion.dtos.BadgeResponse;
import dev.jefy.connectpro.management.domain.repositoty.BadgeRepository;
import dev.jefy.connectpro.marketplace.applicaion.dtos.JobPostListingResponse;
import dev.jefy.connectpro.marketplace.applicaion.dtos.ServiceListingResponse;
import dev.jefy.connectpro.portfolio.applicaion.dtos.JobPostResponse;
import dev.jefy.connectpro.portfolio.applicaion.dtos.PortfolioResponse;
import dev.jefy.connectpro.portfolio.applicaion.dtos.ProjectResponse;
import dev.jefy.connectpro.portfolio.applicaion.dtos.ServiceResponse;
import dev.jefy.connectpro.portfolio.domain.model.JobPost;
import dev.jefy.connectpro.portfolio.domain.model.PService;
import dev.jefy.connectpro.portfolio.domain.model.Portfolio;
import dev.jefy.connectpro.portfolio.domain.repository.JobPostRepository;
import dev.jefy.connectpro.portfolio.domain.repository.PortfolioRepository;
import dev.jefy.connectpro.portfolio.domain.repository.ProjectRepository;
import dev.jefy.connectpro.portfolio.domain.repository.ServiceRepository;
import dev.jefy.connectpro.portfolio.domain.vo.JobPostId;
import dev.jefy.connectpro.portfolio.domain.vo.PortfolioId;
import dev.jefy.connectpro.shared.application.dtos.PortfolioSummaryData;
import dev.jefy.connectpro.shared.application.exceptions.ResourceNotFound;
import lombok.RequiredArgsConstructor;

/**
 * @author Jôph Yamba
 */
@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class PortfolioQueryImpl implements PortfolioQuery {
    private final PortfolioRepository portfolioRepo;
    private final JobPostRepository jobPostRepo;
    private final ServiceRepository serviceRepo;
    private final BadgeRepository badgeRepo;
    private final ProjectRepository projectRepo;
    private final EngagementClient engagementClient;
    private final ManagementClient managementClient;
    
    @Override
    public PortfolioResponse get(PortfolioId portfolioId) {
        Assert.notNull(portfolioId, "id cannot be null");
        var portfolio = getPortfolio(portfolioId);
        
        BadgeResponse badge = portfolio.getBadgeId() == null 
                ? null 
                : badgeRepo.findById(portfolio.getBadgeId())
                .map(BadgeResponse::from)
                .orElse(null);
        
        List<ServiceListingResponse> services = serviceRepo.findAllByPortfolioId(portfolio.getId())
                .stream()
                .map(mapToServiceListingResponse(portfolio))
                .toList();
        
        List<JobPostListingResponse> jobPosts = jobPostRepo.findAllByPortfolioId(portfolio.getId())
                .stream()
                .map(mapToJobPostListingResponse(portfolio))
                .toList();
        
        List<ProjectResponse> projects = projectRepo.findAllByPortfolioId(portfolio.getId())
                .stream()
                .map(ProjectResponse::from)
                .toList();
        
        
        return PortfolioResponse.from(portfolio, badge, services, jobPosts, projects);
    }



    @Override
    public JobPostResponse getJopPost(JobPostId jobPostId) {
        return jobPostRepo.findById(jobPostId)
                .map(mapToJobPostResponse())
                .orElseThrow(()-> new ResourceNotFound("job post with id: " + jobPostId + " not found"));
    }

    @Override
    public ServiceResponse getService(ServiceId serviceId) {
        return serviceRepo.findById(serviceId)
                .map(mapToServiceResponse())
                .orElseThrow(()-> new ResourceNotFound("Service with id " + serviceId + " not found"));
        
    }

    

    private Function<PService, ServiceResponse> mapToServiceResponse() {
        return sevice -> {
            Portfolio portfolio = getPortfolio(sevice.getPortfolioId());
            var category = managementClient.getCategory(sevice.getCategoryId());
            var award = sevice.getAwardId() != null
                    ? managementClient.getAward(sevice.getAwardId()).orElse(null)
                    : null;
            var reviewData = engagementClient.getReviewData(sevice.getId());
            var recentReviews = engagementClient.recentReviews(sevice.getId());
            PortfolioSummaryData portfolioData = PortfolioSummaryData.from(portfolio);
            return ServiceResponse.from(sevice, portfolioData, category, award, reviewData, recentReviews);
        };
    }
    
    private Function<JobPost, JobPostResponse> mapToJobPostResponse() {
        return jobPost -> {
            Portfolio portfolio = getPortfolio(jobPost.getPortfolioId());
            var category = managementClient.getCategory(jobPost.getCategoryId());
            PortfolioSummaryData portfolioData = PortfolioSummaryData.from(portfolio);
            return JobPostResponse.from(jobPost, portfolioData, category);
        };
    }

    private Function<PService, ServiceListingResponse> mapToServiceListingResponse(Portfolio portfolio) {
        return sevice -> {
            var category = managementClient.getCategory(sevice.getCategoryId());
            var award = sevice.getAwardId() != null
                    ? managementClient.getAward(sevice.getAwardId()).orElse(null)
                    : null;
            var reviewData = engagementClient.getReviewData(sevice.getId());
            PortfolioSummaryData portfolioData = PortfolioSummaryData.from(portfolio);
            return ServiceListingResponse.from(sevice, portfolioData, category, award, reviewData);
        };
    }
    private Function<JobPost, JobPostListingResponse> mapToJobPostListingResponse(Portfolio portfolio) {
        return jobPost -> {
            var category = managementClient.getCategory(jobPost.getCategoryId());
            PortfolioSummaryData portfolioData = PortfolioSummaryData.from(portfolio);
            return JobPostListingResponse.from(jobPost, portfolioData, category);
        };
    }

    private @NonNull Portfolio getPortfolio(PortfolioId portfolioId) {
        return portfolioRepo.findById(portfolioId)
                .orElseThrow(() -> new ResourceNotFound("Portfolio with id " + portfolioId + " not found"));
    }
}

File: src/main/kotlin/dev/jefy/connectpro/portfolio/application/query/PortfolioQueryImpl.kt
package dev.jefy.connectpro.portfolio.application.query

import dev.jefy.connectpro.engagement.EngagementClient
import dev.jefy.connectpro.management.ManagementClient
import dev.jefy.connectpro.management.appliacaion.dtos.toResponse
import dev.jefy.connectpro.management.domain.repository.BadgeRepository
import dev.jefy.connectpro.marketplace.MarketplaceClient
import dev.jefy.connectpro.marketplace.application.dtos.JobPostListingResponse
import dev.jefy.connectpro.marketplace.application.dtos.ServiceListingResponse
import dev.jefy.connectpro.marketplace.application.dtos.toListingResponse
import dev.jefy.connectpro.portfolio.application.dtos.JobPostResponse
import dev.jefy.connectpro.portfolio.application.dtos.PortfolioResponse
import dev.jefy.connectpro.portfolio.application.dtos.ServiceResponse
import dev.jefy.connectpro.portfolio.application.dtos.toResponse
import dev.jefy.connectpro.portfolio.application.exceptions.JobPostNotFoundException
import dev.jefy.connectpro.portfolio.application.exceptions.PortfolioNotFoundException
import dev.jefy.connectpro.portfolio.application.exceptions.ServiceNotFoundException
import dev.jefy.connectpro.portfolio.domain.model.JobPost
import dev.jefy.connectpro.portfolio.domain.model.Portfolio
import dev.jefy.connectpro.portfolio.domain.model.Service
import dev.jefy.connectpro.portfolio.domain.repository.JobPostRepository
import dev.jefy.connectpro.portfolio.domain.repository.PortfolioRepository
import dev.jefy.connectpro.portfolio.domain.repository.ProjectRepository
import dev.jefy.connectpro.portfolio.domain.repository.ServiceRepository
import dev.jefy.connectpro.portfolio.domain.vo.JobPostId
import dev.jefy.connectpro.portfolio.domain.vo.PortfolioId
import dev.jefy.connectpro.portfolio.domain.vo.ServiceId
import dev.jefy.connectpro.shared.application.dtos.toSummaryData
import dev.jefy.connectpro.shared.infrastructure.annotations.QueryService
import dev.jefy.connectpro.shared.infrastructure.file_storage.ImageUrlResolver
import org.springframework.util.Assert

@QueryService
class PortfolioQueryImpl(
    private val portfolioRepo: PortfolioRepository,
    private val jobPostRepo: JobPostRepository,
    private val serviceRepo: ServiceRepository,
    private val badgeRepo: BadgeRepository,
    private val projectRepo: ProjectRepository,
    private val engagementClient: EngagementClient,
    private val managementClient: ManagementClient,
    private val marketplaceClient: MarketplaceClient,
    private val resolver: ImageUrlResolver
) : PortfolioQuery {

    override fun get(portfolioId: PortfolioId): PortfolioResponse {
        Assert.notNull(portfolioId, "id cannot be null")
        val portfolio = getPortfolio(portfolioId)

        val badge = portfolio.badgeId?.let { badgeId ->
            badgeRepo.findById(badgeId).map { it.toResponse() }.orElse(null)
        }

        val services = serviceRepo.findAllByPortfolioId(portfolio.id)
            .map{ mapToServiceListingResponse(it, portfolio) }

        val jobPosts = jobPostRepo.findAllByPortfolioId(portfolio.id)
            .map { mapToJobPostListingResponse(it, portfolio) }

        val projects = projectRepo.findAllByPortfolioId(portfolio.id)
            .map{ it.toResponse(resolver)}

        return portfolio.toResponse( badge, services, jobPosts, projects, resolver)
    }

    override fun getJobPost(jobPostId: JobPostId): JobPostResponse {
        return jobPostRepo.findById(jobPostId)
            .map{ mapToJobPostResponse(it)}
            .orElseThrow { JobPostNotFoundException() }
    }

    override fun getAllJobPost(portfolioId: PortfolioId): List<JobPostListingResponse> {
        return jobPostRepo.findAllByPortfolioId(portfolioId)
            .map{ mapToJobPostListingResponse(it, getPortfolio(portfolioId)) }
    }

    override fun getService(serviceId: ServiceId): ServiceResponse {
        return serviceRepo.findById(serviceId)
            .map(mapToServiceResponse())
            .orElseThrow { ServiceNotFoundException() }
    }

    override fun getAllService(portfolioId: PortfolioId): List<ServiceListingResponse> {
        return serviceRepo.findAllByPortfolioId(portfolioId)
            .map{ mapToServiceListingResponse(it, getPortfolio(portfolioId)) }
    }

    private fun mapToServiceResponse(): (Service) -> ServiceResponse = { service ->
        val portfolio = getPortfolio(service.portfolioId)
        val category = managementClient.getCategory(service.categoryId)

        val award = service.awardId?.let { managementClient.getAward(it).orElse(null) }

        val reviewData = engagementClient.getReviewData(service.id)
        val recentReviews = engagementClient.recentReviews(service.id)
        
        service.toResponse(
            portfolio.toSummaryData(resolver),
            category,
            award,
            reviewData,
            recentReviews,
            resolver
        )
    }

    private val mapToJobPostResponse : (JobPost) -> JobPostResponse = { jobPost ->
        val portfolio = getPortfolio(jobPost.portfolioId)
        val category = managementClient.getCategory(jobPost.categoryId)
        
        val userApplied: Boolean = marketplaceClient.hasUserAppliedToJob(jobPostId = jobPost.id)
        jobPost.toResponse(
            portfolio = portfolio.toSummaryData(resolver), 
            category = category, 
            userApplied = userApplied
        )
    }
    
    private val mapToServiceListingResponse: (Service, Portfolio) -> ServiceListingResponse = { service, portfolio ->

        val category = managementClient.getCategory(service.categoryId)

        val award = service.awardId?.let { managementClient.getAward(it).orElse(null) }

        val reviewData = engagementClient.getReviewData(service.id)
        
        service.toListingResponse(
            portfolio.toSummaryData(resolver),
            category,
            award,
            reviewData,
            resolver
        )
    }
    

    private val mapToJobPostListingResponse: (JobPost, Portfolio) -> JobPostListingResponse = { jobPost, portfolio ->
        val category = managementClient.getCategory(jobPost.categoryId)
        val userApplied: Boolean = marketplaceClient.hasUserAppliedToJob(jobPostId = jobPost.id)
        jobPost.toListingResponse(
            portfolio = portfolio.toSummaryData(resolver), 
            category =  category, 
            userApplied = userApplied
        )
    }

    private fun getPortfolio(portfolioId: PortfolioId): Portfolio = portfolioRepo
        .findById(portfolioId)
        .orElseThrow { PortfolioNotFoundException() }
    
}